src/main/sensors/barometer.c

```diff
@@ -475,40 +440,33 @@ uint32_t baroUpdate(timeUs_t currentTimeUs)
     return sleepTime;
 }
 
-static float pressureToAltitude(const float pressure)
-{
-    return (1.0f - powf(pressure / 101325.0f, 0.190295f)) * 4433000.0f;
-}
-
-int32_t baroCalculateAltitude(void)
+float baroCalculateAltitude(void)
 {
-    int32_t BaroAlt_tmp;
-
     // calculates height from ground via baro readings
     if (baroIsCalibrationComplete()) {
-        BaroAlt_tmp = lrintf(pressureToAltitude((float)(baroPressureSum / barometerConfig()->baro_sample_count)));
-        BaroAlt_tmp -= baroGroundAltitude;
-        baro.BaroAlt = lrintf((float)baro.BaroAlt * CONVERT_PARAMETER_TO_FLOAT(barometerConfig()->baro_noise_lpf) + (float)BaroAlt_tmp * (1.0f - CONVERT_PARAMETER_TO_FLOAT(barometerConfig()->baro_noise_lpf))); // additional LPF to reduce baro noise
-    }
-    else {
-        baro.BaroAlt = 0;
+        baro.BaroAlt = filteredBaroAltitudeRaw - baroGroundAltitude;
+    } else {
+        baro.BaroAlt = 0.0f;
     }
+    DEBUG_SET(DEBUG_BARO, 3, baro.BaroAlt); // metres * 100
     return baro.BaroAlt;
 }
 
 void performBaroCalibrationCycle(void)
 {
-    static int32_t savedGroundPressure = 0;
-
-    baroGroundPressure -= baroGroundPressure / 8;
-    baroGroundPressure += baroPressureSum / barometerConfig()->baro_sample_count;
-    baroGroundAltitude = (1.0f - pow_approx((baroGroundPressure / 8) / 101325.0f, 0.190259f)) * 4433000.0f;
-
-    if (baroGroundPressure == savedGroundPressure) {
+    static float savedGroundAltitude = 10000000.0f; // some non-zero value
+
+    baroGroundAltitude = baroGroundAltitude + 0.2f * (filteredBaroAltitudeRaw - baroGroundAltitude);
+    // extra smoothing of baroGroundAltitude during cal period
+    if (fabsf(baroGroundAltitude - savedGroundAltitude) < 0.1f) {
```

Comment:
I tested this for about 20 runs at different values and with different filter strengths.  Because these are floats, and the filter is very slow, we do get very small smooth changes in 25ms.  
What I found was that with these settings, the timeout was reduced but not completely.
A less strong filter or a wider range would lead to the timeout rarely being applied.
Conversely a stronger filter or narrower range had the opposite effect, leading to the timeout usually being reached.
The above settings appeared to return a decent zero point within a shorter time than the timeout.

Reply:
Still sounds suspicious ... 1mm of baro noise means that it is filtered (and thus delayed) too much. And noise in difference is based on baroGroundAltitude , so only main filter is in effect.
IMO this is hit because noise hits 1mm window randomly when values are close enough (much actual difference is orders of magnitude larger)
Can you try to blackbox it ? (probably by restarting calibratingB with BB enabled)

Reply:
Yes I will do that.  
The old method was similar.  A lowpass in addition to the median filter, and then an integer difference of zero would bypass the full `calibratingB` count.
My simple testing was to make the threshold so big that it always ran the full `calibratingB` count.  Then I just reduced it so that sometimes it reduced the time taken :-)
And with this the calibration was OK every time.
The actual value drifts up and down about 0.3-0.5m within a second in sensors tab with default filtering from this PR.
At 1/5th of default this is much slower and about 0.1m, which is only 2.5mm per step.
The initial value of filteredBaroAltitudeRaw is very high. 
Considering both factors, we need a very smoothed, slowly falling lowpass from the large offset, with a slow enough time constant to smooth out most of the random stuff.  Then, I agree, some random stuff superimposed on the average will eventually hit 1.0mm per step, and by this time, typically, the baseline is reasonably stable.
I will log and check.

Reply:
Starting from large value is IMO wrong - a lot of time constants is spent decreasing initial value without providing any useful information. It is much better to use lover cutoff, so that resulting ground estimate is calculated over longer interval (suppressing noise).

As for noise termination - it does work, but termination condition is probability based and hugely different from 1mm in the code. This is misleading and may lead to new problems when someone touches code in future.
IMO it's better to remove it, because correct implementation is not worth the complexity added (I did something roughly similar with light detection and it was really ugly)